LiskHQ/lisk-sdk
# Bus sockets timeout - Closes #3942

### What was the problem?
Bus sockets time out whenever you stop the application (`CMD+C`) and try to restart it.
The reason for this is the path where the sockets are stored is longer than the allowed 104 characters for MacOS (or 108 for Unix). So, the cleanup did happen upon shutting down the application, but because the path is too long, it couldn't clean up those files.

This resulted in having to manually remove the `/tmp` folder.

### How did I fix it?
1. Made the `tempPath` configurable as the user wishes.
2. By default, the path gets created inside the `/tmp` folder of the system.
+ Had to slice the default app `label` as it was using the full payload hash of the genesis block. @lsilvs suggested to use only the first 7 characters as that makes it unique (referring to Github commit slice). 

### How to test it?
Run tests:
- `npm run jest:unit -- test/jest/unit/specs/controller/application.spec.js`
- `framework/test/jest/unit/specs/controller/controller.spec.js`
- Modify the `/config/devnet/config.json` and check if this folder path contains files when the app is running:
```json
{
	"app": {
		"ipc": {
			"enabled": true
		},
		"tempPath": "/tmp/custom-folder"
	},
...
}
```

#### Calculations for determining unix socket path length
Note for future reference about calculations for limiting `tempPath` and `label` config parameters.

Unix socket path length:
- MacOS 104
- Unix 108

We choose to support MacOS and use a buffer of 1 character = 103.
-> Max length app label: 30
-> Max length tempPath: 50

Longest possible path:
`${tempPath:50}/${label:30}/sockets/lisk_pub.sock` = 103 chars 👍 

### Review checklist

* The PR resolves #3942 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated
